core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/bi/GroupBiToBiNode.java

```diff
@@ -0,0 +1,163 @@
+    private final Supplier<ResultContainer_> supplier;
+    private final TriFunction<ResultContainer_, OldA, OldB, Runnable> accumulator;
+    private final Function<ResultContainer_, B> finisher;
+    /**
+     * Calls for example {@link BiScorer#insert(BiTuple)}, {@link JoinTriNode#insertAB(BiTuple)} and/or ...
+     */
+    public final Consumer<BiTuple<A, B>> nextNodesInsert;
+    /**
+     * Calls for example {@link BiScorer#retract(BiTuple)}, {@link JoinTriNode#insertAB(BiTuple)} and/or ...
+     */
+    public final Consumer<BiTuple<A, B>> nextNodesRetract;
+
+    private final Map<BiTuple<OldA, OldB>, GroupPart> groupPartMap;
+    private final Map<A, Group> groupMap;
+    private final Queue<Group> dirtyGroupQueue;
+
+    public GroupBiToBiNode(BiFunction<OldA, OldB, A> groupKeyMapping,
+            BiConstraintCollector<OldA, OldB, ResultContainer_, B> collector,
+            Consumer<BiTuple<A, B>> nextNodesInsert, Consumer<BiTuple<A, B>> nextNodesRetract) {
+        this.groupKeyMapping = groupKeyMapping;
+        supplier = collector.supplier();
+        accumulator = collector.accumulator();
+        finisher = collector.finisher();
+        this.nextNodesInsert = nextNodesInsert;
+        this.nextNodesRetract = nextNodesRetract;
+        groupMap = new HashMap<>(1000);
+        groupPartMap = new HashMap<>(1000);
+        dirtyGroupQueue = new ArrayDeque<>(1000);
+    }
+
+    private final class Group {
+        A groupKey;
+        ResultContainer_ resultContainer;
+        int parentCount = 0;
+        boolean dirty = false;
+        boolean dying = false;
+        BiTuple<A, B> tupleAB = null;
+
+        public Group(A groupKey, ResultContainer_ resultContainer) {
+            this.groupKey = groupKey;
+            this.resultContainer = resultContainer;
+        }
+    }
+
+    private final class GroupPart {
+        Group group;
+        Runnable undoAccumulator;
+
+        public GroupPart(Group group, Runnable undoAccumulator) {
+            this.group = group;
+            this.undoAccumulator = undoAccumulator;
+        }
+    }
+
+    public void insertAB(BiTuple<OldA, OldB> tupleOldAB) {
+        A groupKey = groupKeyMapping.apply(tupleOldAB.factA, tupleOldAB.factB);
+        Group group = groupMap.computeIfAbsent(groupKey, k -> new Group(groupKey, supplier.get()));
+        group.parentCount++;
+
+        Runnable undoAccumulator = accumulator.apply(group.resultContainer, tupleOldAB.factA, tupleOldAB.factB);
+        GroupPart groupPart = new GroupPart(group, undoAccumulator);
+        GroupPart old = groupPartMap.put(tupleOldAB, groupPart);
+        if (old != null) {
+            throw new IllegalStateException("Impossible state: the tuple for the fact ("
+                    + tupleOldAB.factA + ", " + tupleOldAB.factB + ") was already added in the groupPartMap.");
+        }
+        if (!group.dirty) {
+            group.dirty = true;
+            dirtyGroupQueue.add(group);
```

Comment:
Seems redundant. Can we remove one of the two?
And if we can not, can we at least make sure (via some `setDirty()` method) that these two are always called together?

Reply:
We can't remove one, because of performance.
But we could indeed extract a method - I 'll take that along in the next group refactor